src/d/actor/d_a_kui.cpp

```diff
@@ -4,68 +4,541 @@
+        0.0f, 0.0f, 1.0f, 0.0f,
+    };
+    cXyz& eyePos = a_this->eyePos;
+    camera_class* camera = dCam_getCamera();
+
+    cXyz look_dir = eyePos - camera->mLookat.mEye;
+
+    cXyz light_dir;
+    dKyr_get_vectle_calc(&a_this->tevStr.mLightPosWorld, &eyePos, &light_dir);
+
+    cXyz refl;
+    C_VECHalfAngle(&look_dir, &light_dir, &refl);
+    Mtx reflMtx;
+    C_MTXLookAt(reflMtx, &cXyz::Zero, &cXyz::BaseY, &refl);
+
+    mDoMtx_stack_c::scaleS(scale, scale, 1.0f);
+    mDoMtx_stack_c::concat(mtx_adj);
+    mDoMtx_stack_c::concat(reflMtx);
+    MtxP mtx = mDoMtx_stack_c::get();
+    mtx[0][3] = 0.0f;
+    mtx[1][3] = 0.0f;
+    mtx[2][3] = 0.0f;
+
+    Mtx now_copy;
+    PSMTXCopy(mDoMtx_stack_c::get(), now_copy);
+
+    for (u16 i = 0; i < modelData->getMaterialNum(); i++) {
+        J3DMaterial* mat = modelData->getMaterialNodePointer(i);
+        for (u32 j = 0; j < 8; j++) {
+            J3DTexMtx* texMtx = mat->getTexMtx(j);
+            if (texMtx != NULL) {
+                switch (texMtx->getTexMtxInfo().mInfo) {
+                    case 10:
+                    case 11:
+                        texMtx->getTexMtxInfo().setEffectMtx(now_copy);
+                        break;
+                }
+            }
+        }
+    }
 }
 
 /* 0000037C-00000540       .text daKui_Draw__FP9kui_class */
-static BOOL daKui_Draw(kui_class*) {
-    /* Nonmatching */
+static BOOL daKui_Draw(kui_class* i_this) {
+    if (i_this->type == 3) {
+        u32 light_type = 0;
+        if (REG0_S(0) != 0) {
+            light_type = 1;
+        }
+
+        i_this->eyePos = i_this->current.pos;
+        i_this->eyePos.y += REG0_F(10);
+
+        g_env_light.settingTevStruct(light_type, &i_this->eyePos, &i_this->tevStr);
+        g_env_light.setLightTevColorType(i_this->mpModel2, &i_this->tevStr);
+        g_env_light.setLightTevColorType(i_this->mpModel, &i_this->tevStr);
+
+        setEffectMtx(i_this, i_this->mpModel2->getModelData(), REG0_F(11) + 1.0f);
+        setEffectMtx(i_this, i_this->mpModel->getModelData(), REG0_F(12) + 1.0f);
+
+        mDoExt_modelUpdateDL(i_this->mpModel2);
+        mDoExt_modelUpdateDL(i_this->mpModel);
+    } else if (i_this->type != 1) {
+        g_env_light.settingTevStruct(TEV_TYPE_ACTOR, &i_this->current.pos, &i_this->tevStr);
+        g_env_light.setLightTevColorType(i_this->mpModel2, &i_this->tevStr);
+
+        dComIfGd_setListBG();
+        mDoExt_modelUpdateDL(i_this->mpModel2);
+        dComIfGd_setList();
+
+        if (i_this->mpModel) {
+            g_env_light.setLightTevColorType(i_this->mpModel, &i_this->tevStr);
+            mDoExt_modelUpdateDL(i_this->mpModel);
+        }
+    }
+    return TRUE;
 }
 
 /* 00000540-00000920       .text demo_camera__FP9kui_class */
-void demo_camera(kui_class*) {
-    /* Nonmatching */
+static void demo_camera(kui_class* i_this) {
+    camera_class* camera = dComIfGp_getCamera(dComIfGp_getPlayerCameraID(0));
+    s8 bVar2 = true;
+
+    switch ((s8)i_this->field_0x2E8) {
```

Comment:
Since you seem to have a `(s8)` cast on all reads of `field_0x2E8`, that might indicate its type should be changed from `u8` to `s8` and the casts can be removed. Unless you already tried that and it broke something.

Reply:
The cast in the `switch` still seems necessary but the cast at the end of the function goes away changing it to a `s8`.